이 자료는 위키독스 딥 러닝을 이용한 자연어 처리 입문의 네거티브 샘플링 구현하기 튜토리얼입니다.  

링크 : https://wikidocs.net/69141  

2021년 10월 14일에 마지막으로 테스트되었습니다.

## **09-04 네거티브 샘플링을 이용한 Word2Vec 구현(Skip-Gram with Negative Sampling, SGNS)**
--- 
네거티브 샘플링(Negative Sampling)을 사용하는 Word2Vec을 직접 케라스(Keras)를 통해 구현해봅시다.

---
### **1. 네거티브 샘플링(Negative Sampling)**
Word2Vec의 출력층에서는 소프트맥스 함수를 지난 단어 집합 크기의 벡터와 실제값인 원-핫 벡터와의 오차를 구하고 이로부터 임베딩 테이블에 있는 모든 단어에 대한 임베딩 벡터 값을 업데이트합니다. 만약 단어 집합의 크기가 수만 이상에 달한다면 이 작업은 굉장히 무거운 작업이므로, Word2Vec은 꽤나 학습하기에 무거운 모델이 됩니다.

Word2Vec은 역전파 과정에서 모든 단어의 임베딩 벡터값의 업데이트를 수행하지만, 만약 현재 집중하고 있는 중심 단어와 주변 단어가 '강아지'와 '고양이', '귀여운'과 같은 단어라면, 사실 이 단어들과 별 연관 관계가 없는 '돈가스'나 '컴퓨터'와 같은 수많은 단어의 임베딩 벡터값까지 업데이트하는 것은 비효율적입니다.

네거티브 샘플링은 Word2Vec이 학습 과정에서 전체 단어 집합이 아니라 일부 단어 집합에만 집중할 수 있도록 하는 방법입니다. 가령, 현재 집중하고 있는 주변 단어가 '고양이', '귀여운'이라고 해봅시다. 여기에 '돈가스', '컴퓨터', '회의실'과 같은 단어 집합에서 무작위로 선택된 주변 단어가 아닌 단어들을 일부 가져옵니다. 이렇게 하나의 중심 단어에 대해서 전체 단어 집합보다 훨씬 작은 단어 집합을 만들어놓고 마지막 단계를 이진 분류 문제로 변환합니다. 주변 단어들을 긍정(positive), 랜덤으로 샘플링 된 단어들을 부정(negative)으로 레이블링한다면 이진 분류 문제를 위한 데이터셋이 됩니다. 이는 기존의 단어 집합의 크기만큼의 선택지를 두고 다중 클래스 분류 문제를 풀던 Word2Vec보다 훨씬 연산량에서 효율적입니다

---
### **2. 네거티브 샘플링 Skip-Gram(Skip-Gram with Negative Sampling, SGNS)**
앞서 배운 Skip-gram을 상기해봅시다.


![image](https://wikidocs.net/images/page/69141/%EA%B7%B8%EB%A6%BC1.PNG)




Skip-gram은 중심 단어로부터 주변 단어를 예측하는 모델이었습니다. 위와 같은 문장이 있다고 한다면, Skip-gram은 중심 단어 cat으로부터 주변 단어 The, fat, sat, on을 예측합니다. 기존의 Skip-gram 모델을 일종의 주황 박스로 생각해본다면, 아래의 그림과 같이 입력은 중심 단어, 모델의 예측은 주변 단어인 구조입니다.

![image](https://wikidocs.net/images/page/69141/%EA%B7%B8%EB%A6%BC1-1.PNG)


하지만 네거티브 샘플링을 사용하는 Skip-gram(Skip-Gram with Negative Sampling, SGNS) 이하 SGNS는 이와는 다른 접근 방식을 취합니다. SGNS는 다음과 같이 중심 단어와 주변 단어가 모두 입력이 되고, 이 두 단어가 실제로 윈도우 크기 내에 존재하는 이웃 관계인지 그 확률을 예측합니다.

![image](https://wikidocs.net/images/page/69141/%EA%B7%B8%EB%A6%BC1-2.PNG)



기존의 Skip-gram 데이터셋을 SGNS의 데이터셋으로 바꾸는 과정을 봅시다.

![image](https://wikidocs.net/images/page/69141/%EA%B7%B8%EB%A6%BC3.PNG)

위의 그림에서 좌측의 테이블은 기존의 Skip-gram을 학습하기 위한 데이터셋입니다. Skip-gram은 기본적으로 중심 단어를 입력, 주변 단어를 레이블로 합니다. 하지만 SGNS를 학습하고 싶다면, 이 데이터셋을 우측의 테이블과 같이 수정할 필요가 있습니다. 우선, 기존의 Skip-gram 데이터셋에서 중심 단어와 주변 단어를 각각 입력1, 입력2로 둡니다. 이 둘은 실제로 윈도우 크기 내에서 이웃 관계였므로 레이블은 1로 합니다. 이제 레이블이 0인 샘플들을 준비할 차례입니다.

![image](https://wikidocs.net/images/page/69141/%EA%B7%B8%EB%A6%BC4.PNG)



실제로는 입력1(중심 단어)와 주변 단어 관계가 아닌 단어들을 입력2로 삼기 위해서 단어 집합에서 랜덤으로 선택한 단어들을 입력2로 하고, 레이블을 0으로 합니다. 이제 이 데이터셋은 입력1과 입력2가 실제로 윈도우 크기 내에서 이웃 관계인 경우에는 레이블이 1, 아닌 경우에는 레이블이 0인 데이터셋이 됩니다. 그리고 이제 두 개의 임베딩 테이블을 준비합니다. 두 임베딩 테이블은 훈련 데이터의 단어 집합의 크기를 가지므로 크기가 같습니다.

![image](https://wikidocs.net/images/page/69141/%EA%B7%B8%EB%A6%BC5.PNG)

두 테이블 중 하나는 입력 1인 중심 단어의 테이블 룩업을 위한 임베딩 테이블이고, 하나는 입력 2인 주변 단어의 테이블 룩업을 위한 임베딩 테이블입니다. 각 단어는 각 임베딩 테이블을 테이블 룩업하여 임베딩 벡터로 변환됩니다.

![image](https://wikidocs.net/images/page/69141/%EA%B7%B8%EB%A6%BC6.PNG)

각 임베딩 테이블을 통해 테이블 룩업하여 임베딩 벡터로 변환되었다면 그 후의 연산은 매우 간단합니다.

![image](https://wikidocs.net/images/page/69141/%EA%B7%B8%EB%A6%BC7.PNG)

중심 단어와 주변 단어의 내적값을 이 모델의 예측값으로 하고, 레이블과의 오차로부터 역전파하여 중심 단어와 주변 단어의 임베딩 벡터값을 업데이트합니다. 학습 후에는 좌측의 임베딩 행렬을 임베딩 벡터로 사용할 수도 있고, 두 행렬을 더한 후 사용하거나 두 행렬을 연결(concatenate)해서 사용할 수도 있습니다. 아래의 실습에서는 좌측의 행렬을 사용하는 방식을 택했습니다.

---
### **3. 20뉴스그룹 데이터 전처리하기**

In [1]:
import pandas as pd
import numpy as np
import nltk
from nltk.corpus import stopwords
from sklearn.datasets import fetch_20newsgroups
from tensorflow.keras.preprocessing.text import Tokenizer

20뉴스그룹 데이터를 사용합니다. 이번 실습에서는 하나의 샘플에 최소 단어 2개는 있어야 합니다. 그래야만 중심 단어, 주변 단어의 관계가 성립하며 그렇지 않으면 샘플을 구성할 수 없어 에러가 발생합니다. 전처리 과정에서 지속적으로 이를 만족하지 않는 샘플들을 제거하겠습니다.

In [2]:
dataset = fetch_20newsgroups(shuffle=True, random_state=1, remove=('headers', 'footers', 'quotes'))
documents = dataset.data
print('총 샘플 수 :',len(documents))

총 샘플 수 : 11314


총 샘플 수는 11,314개입니다. 전처리를 진행해봅시다. 불필요한 토큰을 제거하고, 소문자화를 통해 정규화를 진행합니다.

In [3]:
news_df = pd.DataFrame({'document':documents})
# 특수 문자 제거
news_df['clean_doc'] = news_df['document'].str.replace("[^a-zA-Z]", " ")
# 길이가 3이하인 단어는 제거 (길이가 짧은 단어 제거)
news_df['clean_doc'] = news_df['clean_doc'].apply(lambda x: ' '.join([w for w in x.split() if len(w)>3]))
# 전체 단어에 대한 소문자 변환
news_df['clean_doc'] = news_df['clean_doc'].apply(lambda x: x.lower())

In [4]:
news_df

document  \
0      Well i'm not sure about the story nad it did s...   
1      \n\n\n\n\n\n\nYeah, do you expect people to re...   
2      Although I realize that principle is not one o...   
3      Notwithstanding all the legitimate fuss about ...   
4      Well, I will have to change the scoring on my ...   
...                                                  ...   
11309  Danny Rubenstein, an Israeli journalist, will ...   
11310                                                 \n   
11311  \nI agree.  Home runs off Clemens are always m...   
11312  I used HP DeskJet with Orange Micros Grappler ...   
11313                                        ^^^^^^\n...   

                                               clean_doc  
0      well sure about story seem biased. what disagr...  
1      yeah, expect people read faq, etc. actually ac...  
2      although realize that principle your strongest...  
3      notwithstanding legitimate fuss about this pro...  
4      well, will have change scoring playoff pool. u...  
...                                                  ...  
11309  danny rubenstein, israeli journalist, will spe...  
11310                                                     
11311  agree. home runs clemens always memorable. kin...  
11312  used deskjet with orange micros grappler syste...  
11313  ^^^^^^ argument with murphy. scared hell when ...  

[11314 rows x 2 columns]

현재 데이터프레임에 Null 값이 있는지 확인합니다.

In [5]:
news_df.isnull().values.any()

False

Null 값이 없지만, 빈 값(empy) 유무도 확인해야 합니다. 모든 빈 값을 Null 값으로 변환하고, 다시 Null 값이 있는지 확인합니다.

In [6]:
news_df[news_df.clean_doc == '']

document clean_doc
51                         
89     \n\n\n\n\n          
150                        
255            \n          
311            \n          
...           ...       ...
11156                      
11158                      
11225          \n          
11284         ...          
11310          \n          

[310 rows x 2 columns]

In [7]:
news_df.replace("", float("NaN"), inplace=True)
news_df

document  \
0      Well i'm not sure about the story nad it did s...   
1      \n\n\n\n\n\n\nYeah, do you expect people to re...   
2      Although I realize that principle is not one o...   
3      Notwithstanding all the legitimate fuss about ...   
4      Well, I will have to change the scoring on my ...   
...                                                  ...   
11309  Danny Rubenstein, an Israeli journalist, will ...   
11310                                                 \n   
11311  \nI agree.  Home runs off Clemens are always m...   
11312  I used HP DeskJet with Orange Micros Grappler ...   
11313                                        ^^^^^^\n...   

                                               clean_doc  
0      well sure about story seem biased. what disagr...  
1      yeah, expect people read faq, etc. actually ac...  
2      although realize that principle your strongest...  
3      notwithstanding legitimate fuss about this pro...  
4      well, will have change scoring playoff pool. u...  
...                                                  ...  
11309  danny rubenstein, israeli journalist, will spe...  
11310                                                NaN  
11311  agree. home runs clemens always memorable. kin...  
11312  used deskjet with orange micros grappler syste...  
11313  ^^^^^^ argument with murphy. scared hell when ...  

[11314 rows x 2 columns]

In [8]:
news_df.isnull().values.any()

True

Null 값이 있음을 확인했습니다. Null 값을 제거합니다.

In [9]:
news_df.dropna(inplace=True)
print('총 샘플 수 :', len(news_df))

총 샘플 수 : 11004


샘플 수가 일부 줄어든 것을 확인할 수 있습니다. NLTK에서 정의한 불용어 리스트를 사용하여 불용어를 제거합니다.

In [10]:
# 불용어 제거
stop_words = stopwords.words('english') # NLTK로부터 불용어를 받아옵니다.
tokenized_doc = news_df['clean_doc'].apply(lambda x: x.split()) # 토큰화

# 불용어를 제거합니다.
tokenized_doc = tokenized_doc.apply(lambda x: [item for item in x if item not in stop_words])

In [11]:
tokenized_doc = tokenized_doc.to_list()

In [12]:
len(tokenized_doc)

11004

불용어를 제거하였으므로 단어의 수가 줄어들었습니다. 모든 샘플 중 단어가 1개 이하인 경우를 모두 찾아 제거하겠습니다.

In [13]:
# 단어가 1개 이하인 샘플의 인덱스를 찾아서 저장하고, 해당 샘플들은 제거.
drop_train = [index for index, sentence in enumerate(tokenized_doc) if len(sentence) <= 1]
print(drop_train)
len(tokenized_doc), len(drop_train)

[44, 353, 486, 1224, 1653, 2323, 2373, 2864, 3292, 3389, 3397, 3398, 3562, 3567, 3595, 3784, 3879, 4185, 4592, 4622, 4952, 4975, 5395, 5530, 6020, 6657, 6728, 6888, 7085, 7961, 8161, 8217, 8288, 8422, 8594, 8627, 9703, 10283, 10447, 10738, 10758, 10904, 10916, 10964]


(11004, 44)

In [14]:
tokenized_doc = [tokenized_doc[i] for i in range(len(tokenized_doc)) if i not in drop_train]
# tokenized_doc = np.delete(tokenized_doc, drop_train, axis=0)  # Error 202408 JH
print('총 샘플 수 :',len(tokenized_doc))


총 샘플 수 : 10960


샘플 수가 다시 줄어들었습니다. 단어 집합을 생성하고, 정수 인코딩을 진행합니다.

In [15]:
tokenizer = Tokenizer()
tokenizer.fit_on_texts(tokenized_doc)

word2idx = tokenizer.word_index
idx2word = {value : key for key, value in word2idx.items()}
encoded = tokenizer.texts_to_sequences(tokenized_doc)

상위 2개의 샘플을 출력해봅시다.

In [16]:
print(encoded[:5])

[[39, 52, 921, 141, 15874, 1675, 541, 275, 865, 12013, 17757, 23990, 29709, 275, 865, 63418, 865, 1121, 1096, 1988, 845, 29710, 907, 726, 20460, 275, 865, 168, 141, 1802, 147, 275, 20461, 17757, 6631, 5696, 75, 63419, 7, 35, 163, 609, 648, 29711, 6897, 23991, 2072, 823, 17758, 1112, 8776, 350, 1065, 15875, 666, 56, 161, 4217, 7192, 1924, 435, 55, 278, 4716, 9261, 2679, 39289], [1276, 424, 3, 51, 6150, 157, 111, 469, 88, 17759, 18, 62, 4717, 2854, 63420, 1035, 397, 39290, 8777, 896, 43, 8314, 312, 13026, 896, 3439, 5909, 528, 18, 86, 4718, 9858, 158, 1395, 119, 149, 5180, 63421, 63422, 17760, 63423, 13026, 897, 63424, 63425, 11157, 17761], [246, 845, 2762, 9262, 4019, 1, 26, 2, 5, 94, 291, 898, 5697, 17, 650, 7, 2538, 63426, 6151, 39291, 307, 30, 13027, 35, 149, 479, 291, 276, 898, 1197, 410, 845, 1, 1210, 898, 63427, 1423, 278, 34, 85, 4577, 39291, 307, 1, 1143, 55, 2, 4321, 738, 308, 150, 11158, 1185, 10461, 651, 15876, 127, 4113, 122, 20462, 14272, 2412], [29712, 1475, 15877, 8315, 1

단어 집합의 크기를 확인합니다.

In [17]:
vocab_size = len(word2idx) + 1 
print('단어 집합의 크기 :', vocab_size)

단어 집합의 크기 : 181822


총 64,277개의 단어가 존재합니다.

---
### **4. 네거티브 샘플링을 통한 데이터셋 구성하기**

토큰화, 정제, 정규화, 불용어 제거, 정수 인코딩까지 일반적인 전처리 과정을 거쳤습니다. 네거티브 샘플링을 통한 데이터셋을 구성할 차례입니다. 이를 위해서는 네거티브 샘플링을 위해서 케라스에서 제공하는 전처리 도구인 skipgrams를 사용합니다. 어떤 전처리가 수행되는지 그 결과를 확인하기 위해서 (꽤 시간이 소요되는 작업이므로) 상위 10개의 뉴스그룹 샘플에 대해서만 수행해봅시다.

In [18]:
from tensorflow.keras.preprocessing.sequence import skipgrams
# 네거티브 샘플링
skip_grams = [skipgrams(sample, vocabulary_size=vocab_size, window_size=10) for sample in encoded[:10]]

결과를 확인합니다. 10개의 뉴스그룹 샘플에 대해서 모두 수행되었지만, 첫번째 뉴스그룹 샘플에 대해서만 확인해보겠습니다.

In [19]:
# 첫번째 샘플인 skip_grams[0] 내 skipgrams로 형성된 데이터셋 확인
pairs, labels = skip_grams[0][0], skip_grams[0][1]
for i in range(5):
    print("({:s} ({:d}), {:s} ({:d})) -> {:d}".format(
          idx2word[pairs[i][0]], pairs[i][0], 
          idx2word[pairs[i][1]], pairs[i][1], 
          labels[i]))

(incidences (29710), europe (1988)) -> 1
(lived (1096), #play. (115526)) -> 0
(europe (1988), :phave (154818)) -> 0
(occured. (20460), 45-5, (124544)) -> 0
(story (921), israels (17757)) -> 1


윈도우 크기 내에서 중심 단어, 주변 단어의 관계를 가지는 경우에는 1의 레이블을 갖도록 하고, 그렇지 않은 경우는 0의 레이블을 가지도록 하여 데이터셋을 구성합니다. 이 과정은 각각의 뉴스그룹 샘플에 대해서 동일한 프로세스로 수행됩니다.

In [20]:
print(len(skip_grams))

10


encoded 중 상위 10개의 뉴스그룹 샘플에 대해서만 수행하였으므로 10이 출력됩니다. 그리고 10개의 뉴스그룹 샘플 각각은 수많은 중심 단어, 주변 단어의 쌍으로 된 샘플들을 갖고 있습니다. 첫번째 뉴스그룹 샘플이 가지고 있는 pairs와 labels의 개수를 출력해봅시다.

In [21]:
# 첫번째 샘플에 대해서 생긴 pairs와 labels
print(len(pairs))
print(len(labels))

2460
2460


이 작업을 모든 뉴스그룹 샘플에 대해서 수행합니다.

In [22]:
## 
skip_grams = [skipgrams(sample, vocabulary_size=vocab_size, window_size=10) for sample in encoded]

---
### **5. Skip-Gram with Negative Sampling(SGNS) 구현하기**

In [23]:
from tensorflow.keras.models import Sequential, Model
from tensorflow.keras.layers import Embedding, Reshape, Activation, Input
from tensorflow.keras.layers import Dot
from tensorflow.keras.utils import plot_model
from IPython.display import SVG

하이퍼파라미터인 임베딩 벡터의 차원은 100으로 정하고, 두 개의 임베딩 층을 추가합니다.

In [24]:
embedding_dim = 100

In [25]:
# 중심 단어를 위한 임베딩 테이블
w_inputs = Input(shape=(1, ), dtype='int32')
word_embedding = Embedding(vocab_size, embedding_dim)(w_inputs)

# 주변 단어를 위한 임베딩 테이블
c_inputs = Input(shape=(1, ), dtype='int32')
context_embedding  = Embedding(vocab_size, embedding_dim)(c_inputs)

각 임베딩 테이블은 중심 단어와 주변 단어 각각을 위한 임베딩 테이블이며 각 단어는 임베딩 테이블을 거쳐서 내적을 수행하고, 내적의 결과는 1 또는 0을 예측하기 위해서 시그모이드 함수를 활성화 함수로 거쳐 최종 예측값을 얻습니다.

In [26]:
dot_product = Dot(axes=2)([word_embedding, context_embedding])
dot_product = Reshape((1,), input_shape=(1, 1))(dot_product)
output = Activation('sigmoid')(dot_product)

c:\Users\jeong\anaconda3-1\envs\py310_yolo\lib\site-packages\keras\src\layers\reshaping\reshape.py:39: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [27]:
model = Model(inputs=[w_inputs, c_inputs], outputs=output)
model.summary()
model.compile(loss='binary_crossentropy', optimizer='adam')
plot_model(model, to_file='skip_gram.png', show_shapes=True, show_layer_names=True, rankdir='TB')

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer         │ (None, 1)         │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ input_layer_1       │ (None, 1)         │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ embedding           │ (None, 1, 100)    │ 18,182,200 │ input_layer[0][0] │
│ (Embedding)         │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ embedding_1         │ (None, 1, 100)    │ 18,182,200 │ input_layer_1[0]… │
│ (Embedding)         │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dot (Dot)           │ (None, 1, 1)      │          0 │ embedding[0][0],  │
│                     │                   │            │ embedding_1[0][0] │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ reshape (Reshape)   │ (None, 1)         │          0 │ dot[0][0]         │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ activation          │ (None, 1)         │          0 │ reshape[0][0]     │
│ (Activation)        │                   │            │                   │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 36,364,400 (138.72 MB)

 Trainable params: 36,364,400 (138.72 MB)

 Non-trainable params: 0 (0.00 B)

You must install pydot (`pip install pydot`) for `plot_model` to work.


모델의 학습은 5에포크 수행하겠습니다.

In [ ]:
### 시간 소요됨
for epoch in range(1, 6):
    loss = 0
    for _, elem in enumerate(skip_grams):
        first_elem = np.array(list(zip(*elem[0]))[0], dtype='int32')
        second_elem = np.array(list(zip(*elem[0]))[1], dtype='int32')
        labels = np.array(elem[1], dtype='int32')
        X = [first_elem, second_elem]
        Y = labels
        loss += model.train_on_batch(X,Y)  
    print('Epoch :',epoch, 'Loss :',loss)

Epoch : 1 Loss : 6223.498048901558
Epoch : 2 Loss : 5555.428535073996


---
### **6. 결과 확인하기**

학습된 모델의 결과를 확인해보겠습니다. 학습된 임베딩 벡터들을 vector.txt에 저장합니다. 그 후 이를 gensim의 models.KeyedVectors.load_word2vec_format()으로 로드하면 쉽게 단어 벡터 간 유사도를 구할 수 있습니다.

In [1]:
import gensim

In [ ]:
f = open('vectors.txt' ,'w')
f.write('{} {}\n'.format(vocab_size-1, embedding_dim))
vectors = model.get_weights()[0]
for word, i in tokenizer.word_index.items():
    f.write('{} {}\n'.format(word, ' '.join(map(str, list(vectors[i, :])))))
f.close()

In [ ]:
w2v = gensim.models.KeyedVectors.load_word2vec_format('./vectors.txt', binary=False)

In [ ]:
# 모델 로드
w2v.most_similar(positive=['disease'])

In [ ]:
w2v.most_similar(positive=['soldiers'])

In [ ]:
w2v.most_similar(positive=['police'])

In [ ]:
w2v.most_similar(positive=['hero'])

In [ ]:
w2v.most_similar(positive=['engine'])

In [ ]:
w2v.most_similar(positive=['doctor'])